soffice: Make selected cell announcement more efficient

michaelweghorn opened the pull request:
### Link to issue number:
Fixes #13232

### Summary of the issue:

Since PR #12849, information about selected cells in LibreOffice Calc is queried using the 'IAccessibleTable2' interface which is
supported from LibreOffice 7.3 on.
The call to the 'selectedCells' method on that interface requests a list of a11y objects for all currently selected cells, of which only the first and the last one are actually needed for the announcement of selected cells.
Since Calc spreadsheets have more than a billion cells, this is inefficient when many cells are selected and resulted in Calc becoming unresponsive.

### Description of how this pull request fixes the issue:

Instead of using the 'selectedCells' method from the 'IAccessibleTable2' interface, the first and last selected cell are now retrieved using the 'accSelection' on the 'IAccessible' object of the table, which avoids that a11y objects for all other selected
cells have to be generated as well.

### Testing strategy:

* use LibreOffice Calc 7.3 or above
* select all cells in LO Calc and check that NVDA announces coordinate + content of the first cell (A1) and the last cell (AMJ1048576) in the spreadsheet
* test a few other selections in the spreadsheet (use shift + arrow keys to increase/decrease selection)

### Known issues with pull request:

none

### Change log entries:
*Changes*

``- Announcement of selected cells for LibreOffice Calc is more efficient and no longer results in a Calc freeze when many cells are selected. (#13232)``

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

michaelweghorn commented:
@leonardder : Any thoughts on this?
Unfortunately, I couldn't see any way to retrieve coordinates of the first and last selected cell without using IAccessibleTable2's `selectedCells` method.

michaelweghorn commented:
> @leonardder : Any thoughts on this? Unfortunately, I couldn't see any way to retrieve coordinates of the first and last selected cell without using IAccessibleTable2's `selectedCells` method.

Had another idea: While I don't see any way to do this via the `IAccessibleTable2` interface, using `IAccessible::get_accSelection` and then the `IEnumVARIANT` to retrieve the first and last selected cell appears to work just fine. I plan to submit an updated PR using that approach.

leonardder commented:
I think your latter idea is great indeed, as long as get_accSelection is stable. I recall it had some buggery in the past though that might have been fixed lately.
I guess it makes sense to only annonce the first and last cell in a selection regardless of count, unless there's only one selected cell of course.

michaelweghorn committed:
soffice: Make selected cell announcement more efficient

Link to issue number:
Fixes #13232

Summary of the issue:
Since commit 32bf54e1f26fec3abf60d81dc79a01e7a1817931
("Improve selection and merged cell announcements in
LibreOffice Calc 7.3 and above (#12849)"), information
about selected cells in LibreOffice Calc is
queried using the 'IAccessibleTable2' interface which is
supported from LibreOffice 7.3 on.
The call to the 'selectedCells' method on that interface
requests a list of a11y objects for all currently selected
cells, of which only the first and the last one are
actually needed for the announcement of selected cells.
Since Calc spreadsheets have more than a billion cells,
this is inefficient when many cells are selected
and resulted in Calc becoming unresponsive.

Description of how this pull request fixes the issue:
Instead of using the 'selectedCells' method from the
'IAccessibleTable2' interface, the first and last
selected cell are now retrieved using the
'accSelection' on the 'IAccessible' object of the table,
which avoids that a11y objects for all other selected
cells have to be generated as well.

michaelweghorn commented:
> Side note: Unrelated to this change, there's an issue that NVDA runs into an exception when only a single cell is selected (e.g. when reducing a previously larger selection to just one cell). I plan to follow up on that next week.

I have now created issue #13276 and PR #13277 for that.

michaelweghorn committed:
soffice: Handle IDispatch returned by IEnumVARIANT

From LibreOffice change
https://gerrit.libreoffice.org/c/core/+/129201 on,
the IEnumVARIANT returns an IDispatch rather than
a child ID, s. a. discussion in
https://github.com/nvaccess/nvda/pull/13277 .

Add a type check and handle both cases.

michaelweghorn commented:
I have added another commit on top that makes this work for the case where the IEnumVARIANT returns an IDispatch instead of a child ID, so this will work with LibreOffice versions including change https://gerrit.libreoffice.org/c/core/+/129201 as well, s. discussion  in PR #13277.

seanbudd commented on source/appModules/soffice.py:
```diff
@@ -208,11 +209,24 @@ def announceSelectionChange(self):
+				tableAccessible = self.table.IAccessibleTable2Object.QueryInterface(IA2.IAccessible2)
+				firstAccessible = tableAccessible.accChild(firstChild).QueryInterface(IA2.IAccessible2)
+				lastAccessible = tableAccessible.accChild(lastChild).QueryInterface(IA2.IAccessible2)
+			else:
```
I think it would be best to capture the type specifically here, and raise an exception if an unexpected type is used.

```suggestion
			elif isinstance(firstChild, IDispatch):  # from comtypes.automation import IDispatch
				...
			else:
				raise RuntimeError(f"Unexpected LibreOffice object {firstChild}, type: {type(firstChild)}")
```

michaelweghorn replied:
That sounds reasonable. I have added a commit on top doing so. The exact type used here is comtypes.client.dynamic._Dispatch, not comtypes.automation.IDispatch, so that commit uses this.

seanbudd commented on source/appModules/soffice.py:
```diff
@@ -208,11 +209,24 @@ def announceSelectionChange(self):
+			count = self.table.IAccessibleTable2Object.nSelectedCells
+			selection = self.table.IAccessibleObject.accSelection
+			enumObj = selection.QueryInterface(oleacc.IEnumVARIANT)
+			firstChild, _retrievedCount = enumObj.Next(1)
```
```suggestion
			(firstChild, _retrievedCount): Tuple[Union[int, IDispatch], int] = enumObj.Next(1)
```

seanbudd replied:
My mistake - it appears you need to add types before unpacking a tuple according to [PEP](https://peps.python.org/pep-0526/#global-and-local-variable-annotations).

Example
```suggestion
			firstChild: Union[int, IDispatch]
			firstChild, _retrievedCount = enumObj.Next(1)
```

seanbudd commented on source/appModules/soffice.py:
```diff
@@ -208,11 +209,24 @@ def announceSelectionChange(self):
+			firstChild, _retrievedCount = enumObj.Next(1)
+			# skip over all except the last element
+			enumObj.Skip(count - 2)
+			lastChild, _retrieveCount = enumObj.Next(1)
```
```suggestion
			(lastChild, _retrievedCount): Tuple[Union[int, IDispatch], int]  = enumObj.Next(1)
```

michaelweghorn replied:
Same here as for the previous comment: I get an exception when applying this.

seanbudd replied:
```suggestion
			lastChild: Union[int, comtypes.client.dynamic._Dispatch]
			lastChild, _retrieveCount = enumObj.Next(1)
```

seanbudd reviewed (dismissed):
Apologies for the late review @michaelweghorn.
The approach generally looks good, with some suggestion on handling the types.
I've also fixed up the PR description.

michaelweghorn committed:
Apply suggestion wrt type from review

s. https://github.com/nvaccess/nvda/pull/13233#discussion_r909331023

> I think it would be best to capture the type specifically here, and raise an exception if an unexpected type is used.

seanbudd committed:
add type hints to children

seanbudd reviewed (approved):
Thanks @michaelweghorn

seanbudd committed:
Merge branch 'master' into michaelweghorn/issue13232_limit_queried_cells

seanbudd committed:
update changes

seanbudd committed:
import union

michaelweghorn commented:
Thanks @seanbudd - also for helping out with my limited Python skills.
